src/promptflow-devkit/promptflow/_sdk/_service/apis/collector.py

```diff
@@ -49,50 +49,81 @@ def trace_collector(
     content_type = request.headers.get("Content-Type")
     # binary protobuf encoding
     if "application/x-protobuf" in content_type:
-        traces_request = ExportTraceServiceRequest()
-        traces_request.ParseFromString(request.data)
-        all_spans = []
-        for resource_span in traces_request.resource_spans:
-            resource_attributes = dict()
-            for attribute in resource_span.resource.attributes:
-                attribute_dict = json.loads(MessageToJson(attribute))
-                attr_key, attr_value = parse_kv_from_pb_attribute(attribute_dict)
-                resource_attributes[attr_key] = attr_value
-            if SpanResourceAttributesFieldName.COLLECTION not in resource_attributes:
-                resource_attributes[SpanResourceAttributesFieldName.COLLECTION] = TRACE_DEFAULT_COLLECTION
-            resource = {
-                SpanResourceFieldName.ATTRIBUTES: resource_attributes,
-                SpanResourceFieldName.SCHEMA_URL: resource_span.schema_url,
-            }
-            for scope_span in resource_span.scope_spans:
-                for span in scope_span.spans:
-                    # TODO: persist with batch
-                    span: Span = TraceOperations._parse_protobuf_span(span, resource=resource, logger=logger)
-                    if not cloud_trace_only:
-                        all_spans.append(copy.deepcopy(span))
-                        span._persist()
-                        logger.debug("Persisted trace id: %s, span id: %s", span.trace_id, span.span_id)
-                    else:
-                        all_spans.append(span)
-
-        if cloud_trace_only:
-            # If we only trace to cloud, we should make sure the data writing is success before return.
-            _try_write_trace_to_cosmosdb(
-                all_spans, get_created_by_info_with_cache, logger, credential, is_cloud_trace=True
-            )
-        else:
-            # Create a new thread to write trace to cosmosdb to avoid blocking the main thread
-            ThreadWithContextVars(
-                target=_try_write_trace_to_cosmosdb,
-                args=(all_spans, get_created_by_info_with_cache, logger, credential, False),
-            ).start()
+        trace_request = ExportTraceServiceRequest()
+        trace_request.ParseFromString(request.data)
+        collect_trace_request(
+            trace_request=trace_request,
+            get_created_by_info_with_cache=get_created_by_info_with_cache,
+            logger=logger,
+            cloud_trace_only=cloud_trace_only,
+            credential=credential,
+        )
         return "Traces received", 200
 
     # JSON protobuf encoding
     elif "application/json" in content_type:
         raise NotImplementedError
 
 
+def collect_trace_request(
```

Comment:
as offline sync, it's kind of wired to import this; i will move this inside this pr